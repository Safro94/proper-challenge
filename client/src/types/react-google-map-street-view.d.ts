declare module 'react-google-map-street-view';
