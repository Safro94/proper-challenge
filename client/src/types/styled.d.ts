import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				black: string;
				gray: string;
				white: string;
				danger: string;
			};
			primary: IPalette;
		};
	}
}
