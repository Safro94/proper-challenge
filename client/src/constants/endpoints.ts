export const PROPERTIES_ENDPOINT: string = '/api/properties';
