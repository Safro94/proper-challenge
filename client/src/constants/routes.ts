export const HOME = '/';
export const PROPERTY_DETAIL = '/property/:propertyId';
export const NEW_PROPERTY = '/property/add';
