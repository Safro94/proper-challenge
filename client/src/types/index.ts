export interface IProperty {
	id: number;
	address: string;
	size: string;
	rooms: number;
	utilities: string;
	tenantName: string;
}

export enum ButtonTypes {
	Primary = 'primary',
	Danger = 'danger',
}

export enum RequestStatus {
	Idle = 'idle',
	Pending = 'pending',
	Resolved = 'resolved',
	Rejected = 'rejected',
}
