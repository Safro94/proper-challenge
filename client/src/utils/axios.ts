import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);
instance.defaults.headers.common['x-api-key'] = process.env.REACT_APP_API_KEY;

export default instance;
