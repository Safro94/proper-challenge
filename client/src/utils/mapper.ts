import { IAddress, IDawaResponse } from '../types';

const mapDawaResponse = (data: IDawaResponse[]): IAddress[] => {
	return data.map(item => ({
		text: item.tekst,
		type: item.type,
	}));
};

export { mapDawaResponse };
