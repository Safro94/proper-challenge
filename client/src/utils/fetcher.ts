import { RequestMethods, RequestStatus } from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
	headers?: null | any;
}

interface IResponse {
	data: any;
	status: RequestStatus;
}

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
	headers = null,
}: IFetcher): Promise<IResponse> => {
	let data;
	let status = RequestStatus.Pending;

	try {
		const result = await axios[method](
			url,
			JSON.parse(headers),
			JSON.parse(body)
		);
		data = result.data;
		status = RequestStatus.Resolved;
	} catch (error) {
		status = RequestStatus.Rejected;
		throw error;
	}

	return { data, status };
};

export default fetcher;
