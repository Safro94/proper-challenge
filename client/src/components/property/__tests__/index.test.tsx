import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '../../../utils/test-utils';

import Property from '..';

import fetcher from '../../../utils/fetcher';

import { PROPERTY_DETAIL } from '../../../constants/routes';
import { PROPERTIES_ENDPOINT } from '../../../constants/endpoints';
import { RequestMethods } from '../../../types';

jest.mock('../../../utils/fetcher');

describe('Property', () => {
	const index = 1;
	const getItemProps = () => {};
	const property = {
		id: 1,
		tenantName: 'Matias',
		rooms: 3,
		size: '40mt2',
		address: 'abc',
		utilities: 'def',
	};

	it('should render correctly', () => {
		const { container } = render(
			<Property item={property} index={index} getItemProps={getItemProps} />
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	describe('Property functionality', () => {
		const { reload } = window.location;

		beforeAll(() => {
			Object.defineProperty(window, 'location', {
				writable: true,
				value: { reload: jest.fn() },
			});
		});

		afterAll(() => {
			window.location.reload = reload;
		});

		beforeEach(() => {
			render(
				<Property item={property} index={index} getItemProps={getItemProps} />
			);

			(fetcher as jest.Mock).mockImplementation(() => Promise.resolve());
		});

		it('should have a link with the correct url', () => {
			expect(screen.getByRole('link')).toBeInTheDocument();
			expect((screen.getByRole('link') as HTMLLinkElement).href).toBe(
				`http://localhost${PROPERTY_DETAIL.replace(
					':propertyId',
					`${property.id}`
				)}`
			);
		});

		it('should call handleDeleteProperty, call the fetcher and reload the window', async () => {
			const button = screen.getByRole('button');
			expect(button).toBeInTheDocument();
			expect(fetcher).not.toHaveBeenCalled();

			userEvent.click(button);

			await waitFor(() => {
				expect(fetcher).toHaveBeenCalledTimes(1);
				expect(fetcher).toHaveBeenCalledWith({
					url: `${process.env.REACT_APP_SERVER_URL}${PROPERTIES_ENDPOINT}/${property.id}`,
					method: RequestMethods.DELETE,
				});

				expect(window.location.reload).toHaveBeenCalledTimes(1);
			});
		});
	});
});
