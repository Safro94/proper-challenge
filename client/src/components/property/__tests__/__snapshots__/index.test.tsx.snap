// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property should render correctly 1`] = `
<li
  class="sc-jSFjdj eDBPpq"
>
  <div
    class="sc-gKAaRy klXkns"
  >
    <div
      class="sc-fujyAs iWqvxR"
    >
      <img
        alt="Property 1"
        class="sc-pNWdM bhKqTs"
        src="house.jpeg"
      />
    </div>
    <div
      class="sc-gtsrHT dBklPe"
    >
      <div
        class="sc-dlnjwi enLziH"
      >
        <h4
          class="sc-hKFxyN"
        >
          tenant
        </h4>
        <h5
          class="sc-eCApnc daHCGb"
        >
          Matias
        </h5>
      </div>
      <div
        class="sc-dlnjwi enLziH"
      >
        <h4
          class="sc-hKFxyN"
        >
          rooms
        </h4>
        <h5
          class="sc-eCApnc daHCGb"
        >
          3
        </h5>
      </div>
      <div
        class="sc-dlnjwi enLziH"
      >
        <h4
          class="sc-hKFxyN"
        >
          size
        </h4>
        <h5
          class="sc-eCApnc daHCGb"
        >
          40mt2
        </h5>
      </div>
      <div
        class="sc-dlnjwi enLziH"
      >
        <h4
          class="sc-hKFxyN"
        >
          address
        </h4>
        <h5
          class="sc-eCApnc daHCGb"
        >
          abc
        </h5>
      </div>
      <div
        class="sc-dlnjwi enLziH"
      >
        <h4
          class="sc-hKFxyN"
        >
          utilities
        </h4>
        <h5
          class="sc-eCApnc daHCGb"
        >
          def
        </h5>
      </div>
    </div>
    <div
      class="sc-iCoGMd iitrTY"
    >
      <a
        class="sc-jrsJWt gwtpa"
        href="/property/1"
      >
        viewMore
      </a>
      <button
        class="sc-bdnxRM jjtgId"
        type="button"
      >
        deleteProperty
      </button>
    </div>
  </div>
</li>
`;
